### PR TITLE
fix: do not clear edge activity state when session moved to another node

### PR DESCRIPTION
**Pull request by volodymyr-babak**
## Problem

`EdgeGrpcService.onEdgeDisconnect` persists `active=false` based only on **this node's** in-memory session map. In a multi-node TB-Core cluster:

1. Edge drops on node A and reconnects to node B.
2. Node B's `onEdgeConnect` sets `active=true` and claims the edge in the cluster-wide service-id cache.
3. Node A eventually notices its dead stream, finds the session still current in its own local map, and writes `active=false`.

If A's write lands last, the edge stays marked inactive while its gRPC session is alive and uplink telemetry keeps flowing.

That flag is not just a status indicator. `BaseEdgeProcessor.saveEdgeEvent()` reads it before queueing any server→edge event, and actions outside the `doSaveIfEdgeIsOffline` allowlist — device and entity-group assignments among them — are **dropped outright** rather than queued for later delivery. So a stale `active=false` makes assignments silently never reach the edge, and they are not replayed on reconnect. A full *Sync Edge* is the only recovery, because the sync path never consults the attribute.

Reported in a production Kubernetes deployment: assigning a device group to an edge did nothing, unassigning did nothing, *Sync Edge* worked every time, and only an edge restart restored normal operation.

## Fix

Guard the write with the cluster-wide service-id cache and skip it only when **another** node is the recorded owner:

```java
if (!isOwnedByAnotherNode(edgeId)) {
    save(tenantId, edgeId, ACTIVITY_STATE, false);
}
```

The predicate deliberately does **not** reuse the existing `isOwnedByThisNode`. An absent cache entry (expired, or never written) means ownership is *unknown*, not that another node owns it — inverting through `isOwnedByThisNode` would skip the write in that case and leave a genuinely disconnected edge marked active forever. `isOwnedByAnotherNode` fails safe toward the current behaviour.

## Deliberately out of scope

- `LAST_DISCONNECT_TIME` and the rule-engine `DISCONNECT_EVENT` stay unguarded — they are accurate statements about this session, and the connect/disconnect timestamps are the primary diagnostic signal for flapping edges.
- The `saveEdgeEvent` drop policy itself is untouched. Making assignment events queue while an edge is offline is a separate, platform-wide behaviour change and does not belong in an LTS patch.

## Testing

Three cases added to the existing `EdgeGrpcServiceTest`, driving the real `onEdgeDisconnect` path and asserting on the actual `saveAttributes` call rather than the predicate in isolation:

- owned by another node → inactive state **not** persisted
- owned by this node → persisted
- no cache entry → persisted (the fail-safe)

Verified red before green: with the guard reverted, only `givenEdgeReconnectedToAnotherNode_whenDisconnect_thenInactiveStateNotPersisted` fails while the two fail-safe cases still pass.

```
EdgeGrpcServiceTest:  Tests run: 11, Failures: 0, Errors: 0, Skipped: 0
Edge*Test (module):   Tests run: 78, Failures: 0, Errors: 0, Skipped: 1
```

## Follow-up

Same change is needed on `lts-4.2` (which has the service-id cache but not the `isOwnedBy*` helpers) and on the corresponding PE branches.

### CODE REVIEW

File: application/src/main/java/org/thingsboard/server/service/edge/rpc/EdgeGrpcService.java [hunk 1]
```diff
@@ -580,6 +586,13 @@ private boolean isOwnedByThisNode(EdgeId edgeId) {
         return wrapper != null && serviceInfoProvider.getServiceId().equals(wrapper.get());
     }
 
+    // The edge's service-id cache entry points at a different node, i.e. another node is the recorded owner.
+    // An absent entry means ownership is unknown, not that another node owns it.
+    private boolean isOwnedByAnotherNode(EdgeId edgeId) {
+        TbCacheValueWrapper<String> wrapper = edgeIdServiceIdCache.get(edgeId);
+        return wrapper != null && !serviceInfoProvider.getServiceId().equals(wrapper.get());
```
Comment (volodymyr-babak) [inline application/src/main/java/org/thingsboard/server/service/edge/rpc/EdgeGrpcService.java]: This predicate treats "cache entry names some other service id" as "that node is live and owns the edge", but nothing guarantees the recorded owner is still alive. `CACHE_SPECS_EDGE_SESSIONS_TTL` defaults to `0` in `thingsboard.yml` (persistent for the Redis impl), and the only evict point is `evictServiceIdCacheIfOwnedByThisNode`, which only the owning node ever reaches.

So consider: edge E is on A (cache=A, `active=true`); E flaps to B, which writes `active=true` and cache=B while A still holds a stale session; B is then SIGKILLed / OOM-killed, so it never evicts and never writes `active=false`; E stays down. When A finally notices its stale session, `isOwnedByAnotherNode` returns true and A skips the write, so `ACTIVITY_STATE` stays `true` forever — `onEdgeConnect` and `onEdgeDisconnect` here are the only writers of that attribute, and there's no reconciler to correct it. Pre-PR that same sequence happened to self-heal because A wrote `false`.

Since failing safe toward `false` is explicitly the design goal for the unknown-ownership case, is it worth applying the same reasoning to the dead-owner case — e.g. also requiring `wrapper.get()` to be in `partitionService.getAllServiceIds(ServiceType.TB_CORE)` (`DefaultTbClusterService.broadcastEdgeNotification` already uses that), or giving the edge sessions cache a non-zero TTL so a crashed owner's claim expires?

**Review by volodymyr-babak — COMMENTED** [review]
## Review summary

Reviewed 2 changed files in [fix: do not clear edge activity state when session moved to another node](https://github.com/thingsboard/thingsboard/pull/16155). Left 1 comment inline.

The guard is sound for the flapping-edge case it targets and the three added tests pin it properly. The main point worth a look is that `isOwnedByAnotherNode` treats a cache entry naming another service id as "that node owns a live session", with no liveness check — if the recorded owner dies without evicting, the attribute can stay `active=true` indefinitely, which is the mirror image of the bug being fixed. Two adjacent race windows in `onEdgeConnect` / `onEdgeDisconnect` are noted below.

## Additional findings

These observations are about existing code outside the PR's diff — spotted while reading surrounding context.

- **application/src/main/java/org/thingsboard/server/service/edge/rpc/EdgeGrpcService.java:381** — The signal the new guard relies on is published later than the state it is meant to protect: `onEdgeConnect` does `save(ACTIVITY_STATE, true)` at line 378 and only puts the cache entry at line 381. So node B can already have submitted `active=true` while the cache still says A (or nothing), and if A's `onEdgeDisconnect` reads the cache in that window it sees "not another node" and submits `active=false` afterwards — the exact end state this PR is trying to prevent. The stale-session case the PR targets has a much wider window, so the fix does handle the common case, but publishing ownership first (moving the `edgeIdServiceIdCache.put` above the `ACTIVITY_STATE` write) would close most of the remaining hole for free. Worth noting the check-then-write still isn't atomic across nodes even then, since the two `saveAttributes` calls are async and issued from different nodes with no ordering guarantee.
- **application/src/main/java/org/thingsboard/server/service/edge/rpc/EdgeGrpcService.java:529** — The new guard covers the cross-node variant of "stale session clobbers live state", but the same-node variant a few lines above is still open and the guard can't catch it, because there the cache legitimately points at this node. `sessions.get(edgeId)` at line 527 and `sessions.remove(edgeId)` at line 529 are not atomic: if the old session's close callback (one gRPC stream thread) reads `current == session1` and passes the id check, and `onEdgeConnect` for session2 (another gRPC stream thread) does its `sessions.put` in between, the unconditional remove drops the *new* live session2 out of the map, `destroySession(toRemove)` tears down session2's stream, and `ACTIVITY_STATE` is set to false while the edge is connected to this very node — the identical symptom, including the silent `saveEdgeEvent` drops. `sessions.remove(edgeId, current)` (returning early when it returns false) would make this branch claim-once the same way `fireDelayedDisconnectNotification`'s identity-keyed remove already does. Pre-existing rather than introduced here, but it's the same failure mode the PR is closing, so it may be worth folding in.

---
*This review was auto-generated. Findings may contain errors — please verify before applying changes.*

File: application/src/main/java/org/thingsboard/server/service/edge/rpc/EdgeGrpcService.java (hunk 1 above)
Comment (volodymyr-babak) [reply application/src/main/java/org/thingsboard/server/service/edge/rpc/EdgeGrpcService.java]: Good catch — this is a regression I introduced, and the scenario is exactly right. Verified the premises: `CACHE_SPECS_EDGE_SESSIONS_TTL` is `0` (thingsboard.yml:733), and `evictServiceIdCacheIfOwnedByThisNode` is the only evict point in the codebase, reachable only by the owning node. So a SIGKILLed owner's claim is permanent, and since `onEdgeConnect`/`onEdgeDisconnect` are the only writers of `ACTIVITY_STATE`, nothing would ever correct it.

Fixed by requiring the recorded owner to still be a live TB-Core service, per your first suggestion:

```java
return partitionService.getAllServiceIds(ServiceType.TB_CORE).contains(ownerServiceId);
```

Preferred that over a cache TTL because the TTL would also change routing behaviour in `DefaultTbClusterService.processEdgeNotification`, which falls back to a broadcast when the entry is missing — a wider blast radius than this fix warrants on an LTS branch.

One residual trade-off worth naming: during a discovery blip the live owner could briefly be absent from `getAllServiceIds`, and we would then clear the flag — i.e. the original bug, transiently. I took that deliberately, on the grounds that it requires discovery instability *and* a concurrent stale-session disconnect, whereas the dead-owner case needs only a single `kill -9`. Failing toward `false` also keeps the pre-PR outcome in every ambiguous case, which was the design goal you quoted.

Covered by `givenOwnerNodeNoLongerInCluster_whenDisconnect_thenInactiveStateIsPersisted`; confirmed it fails when the liveness check is removed.

**Comment by volodymyr-babak** [comment]
Thanks — all three findings were valid and are now addressed in bcf914d. Verified each against the code before acting rather than taking them on faith; notes below.

### 1. Dead owner honoured indefinitely (inline)

Valid, and a regression introduced by this PR. Answered in detail on the inline thread. Now requires the recorded owner to still be in `partitionService.getAllServiceIds(ServiceType.TB_CORE)`; chose that over a cache TTL because a TTL would also alter the routing fallback in `DefaultTbClusterService.processEdgeNotification`.

### 2. Ownership published after the state it protects (line 381)

Valid. Confirmed the ordering — `save(ACTIVITY_STATE, true)` at 378, `edgeIdServiceIdCache.put` at 381 — so the guard could read a stale owner in exactly the window it is meant to cover. The `put` now happens first:

```java
edgeIdServiceIdCache.put(edgeId, serviceInfoProvider.getServiceId());
save(tenantId, edgeId, ACTIVITY_STATE, true);
```

Safe to move earlier because `sessions.put` already happened above it, so this node is a valid routing target by that point. Agreed the check-then-write still is not atomic across nodes — two async `saveAttributes` from different nodes have no ordering guarantee. This narrows the window rather than closing it, and a genuine fix would need the write to be conditional at the storage layer. Not attempting that on an LTS branch.

### 3. Same-node `sessions.get` / `sessions.remove` race (line 529)

Valid, and I have fixed it here despite it being pre-existing. It produces the identical customer-visible symptom — live session torn down, `active=false` while connected, assignment events silently dropped — and the cross-node guard provably cannot catch it, since the cache legitimately points at this node. Leaving it would have meant shipping a fix whose stated purpose is "a stale session must not clobber live state" while one variant of exactly that remained open.

Used the claim-once idiom already in the file:

```java
if (current != null && current.getSessionId().equals(sessionId) && sessions.remove(edgeId, current)) {
```

Losing the CAS falls through to the stale-session branch, which removes by session id and tears down the stale session — the correct handling for a callback that no longer owns anything.

### Tests

Two added, both verified red against their own fix before going green:

- `givenOwnerNodeNoLongerInCluster_whenDisconnect_thenInactiveStateIsPersisted`
- `givenSessionReplacedDuringDisconnect_whenStaleCallbackRuns_thenLiveSessionKept` — drives the interleaving deterministically by swapping the map entry from within the `getSessionId()` stub, so no concurrency harness is needed

```
EdgeGrpcServiceTest:  Tests run: 13, Failures: 0, Errors: 0, Skipped: 0
Edge*Test (module):   Tests run: 80, Failures: 0, Errors: 0, Skipped: 1
```

**Comment by volodymyr-babak** [comment]
## Merge chain — order matters

Rebuilt so the fix propagates along the real merge paths instead of four independent implementations. Each downstream branch carries the upstream fix commit as a genuine ancestor:

```
CE lts-4.2  thingsboard#16156          <- the change itself
  |
  +-> CE lts-4.3  thingsboard#16155     (merge + 4.3 adaptation)
  |     |
  |     +-> PE lts-4.3  thingsboard-pe#5130   (clean merge from CE)
  |
  +-> PE lts-4.2  thingsboard-pe#5131   (clean merge from CE)
```

**Please merge in this order:** CE `lts-4.2` -> CE `lts-4.3` -> PE `lts-4.2` -> PE `lts-4.3`.

Verified after rebuilding: CE `lts-4.2` is already an ancestor of CE `lts-4.3`, and the CE branches are already ancestors of their PE counterparts, so each of those merges is a fast-forward no-op. PE `lts-4.2` -> PE `lts-4.3` was trial-merged and resolves with zero conflicts.

The earlier revisions of these branches were independent patches per line, which conflicted in `EdgeGrpcService.java` and add/add-conflicted on `EdgeGrpcServiceTest.java` the moment `lts-4.2` was merged into `lts-4.3`. Force-pushed; file contents are byte-identical to what was reviewed.

**Comment by volodymyr-babak** [comment]
## Review follow-ups applied across all branches

All findings were valid, verified in code before acting. Details on the inline threads; summary here.

| # | Finding | Action |
|---|---|---|
| 1 | Guard missed a replacement landing on **this** node | Fixed — added the local-session check, matching the eviction guard |
| 2 | `LAST_DISCONNECT_TIME` + `DISCONNECT_EVENT` ran unconditionally; event payload is `{"active": false}` with `SERVER_SCOPE` | Fixed — both moved inside the guard; overturns the "out of scope" call in the original description |
| 3 | `onEdgeConnect` publish order unpinned by any test | Fixed — `InOrder` assertion added |
| 4 | Guard narrows rather than closes the race; dead-owner comment overclaimed | Comment corrected; callback re-assert declined with rationale on the thread |
| 5 | Other unconditional `removeByEdgeId` has the same hazard (outside the diff) | Fixed — uses the new compare-and-remove |

Finding 2 is the one that changed a decision rather than adding a check: the `DISCONNECT_EVENT` payload is exactly what a Save Attributes node persists, so a tenant rule chain wiring Disconnect Event to Save Attributes would write `active=false` straight back through the rule engine. My original rationale defended the diagnostic value of the timestamps and said nothing about the payload being persisted as the attribute.

Finding 5 is a scope extension beyond what was asked — it is a one-liner using the primitive this change already introduces, and leaving it would have undercut the invariant the PR establishes. Say the word if you would rather it moved to its own PR.

### Propagated along the merge chain

Findings 1, 2, 3 and 4 applied to the 4.2/4.3 branches too, since they share the same gaps — implemented on CE `lts-4.2` and merged down, so `lts-4.2` stays mergeable into `lts-4.3` and CE into PE:

```
CE lts-4.2 (#16156) -> CE lts-4.3 (#16155) -> PE lts-4.3 (pe#5130)
                    \-> PE lts-4.2 (pe#5131)
CE lts-4.4 (#16160) -> PE lts-4.4 (pe#5132)
```

Re-verified after pushing: every CE branch is still an ancestor of its PE counterpart, and PE `lts-4.2` into PE `lts-4.3` trial-merges with zero conflicts.

### Tests

| Branch | Tests |
|---|---|
| CE lts-4.2 | 11 (58 across `Edge*Test`) |
| CE lts-4.3 | 17 |
| PE lts-4.2 | 11 |
| PE lts-4.3 | 17 |
| CE / PE lts-4.4 | 9 each |

Every new test verified red against its own fix before going green.

One note worth recording: on lts-4.2 the `DISCONNECT_EVENT` suppression assertion was initially **vacuous** — an unstubbed `ctx.getRuleProcessor()` NPE'd inside the try/catch before the push could happen, so `never()` passed with or without the guard. Caught it during the red check. It now has a positive counterpart test plus a lenient stub, so the suppression is genuinely asserted.
